peer/peer.go

```diff
@@ -89,38 +93,99 @@ func New(ip string, port int) (*Peer, error) {
+// opens a new stream with this peer.
+// This should be passed as the second argument to SetStreamHandler() after this
+// peer is initialized.
 func (p *Peer) Receive(s net.Stream) {
-	log.Debug("Setting basic stream handler.")
-	p.doCumulus(s)
-}
+	defer s.Close()
+
+	// Get remote peer's full multiaddress
+	remoteMA, err := makeMultiaddr(
+		s.Conn().RemoteMultiaddr(), s.Conn().RemotePeer())
+	if err != nil {
+		log.Fatal("Failed to obtain valid remote peer multiaddress")
+	}
+
+	// Add the remote peer to this peer's subnet
+	err = p.subnet.AddPeer(remoteMA, s)
+	if err != nil {
+		// Subnet is full, advertise other available peers and then close
+		// the stream
+		log.Debug("Peer subnet full. Advertising peers...")
+		p.advertisePeers(s)
+		return
+	}
+	defer p.subnet.RemovePeer(remoteMA)
 
-// Communicate with peers.
-// TODO: Update this to do something useful. For now it just reads from the
-// stream and writes back what it read.
-func (p *Peer) doCumulus(s net.Stream) {
 	buf := bufio.NewReader(s)
-	str, err := buf.ReadString('\n')
+	strMsg, err := buf.ReadString('\n') // TODO: set timeout here
 	if err != nil {
 		log.Error(err)
 		return
 	}
 
-	log.Debugf("Peer %s read: %s", p.ID(), str)
-
-	_, err = s.Write([]byte(str))
+	// Turn the string into a message we can deal with
+	message, err := msg.FromString(strMsg)
 	if err != nil {
 		log.Error(err)
+		return
+	}
+
+	log.Debugf("Peer %s message:\n%s", p.ID(), strMsg)
+
+	// Respond to message
+	p.handleMessage(*message, s)
+}
+
+// Connect adds the given multiaddress to p's Peerstore and opens a stream
+// with the peer at that multiaddress if the multiaddress is valid, otherwise
+// returns error. On success the stream and corresponding multiaddress are
+// added to this peer's subnet.
+func (p *Peer) Connect(peerma string) (net.Stream, error) {
+	peerid, targetAddr, err := extractPeerInfo(peerma)
+	if err != nil {
+		return nil, err
+	}
+
+	// Store the peer's address in this host's PeerStore
+	p.Peerstore().AddAddr(peerid, targetAddr, pstore.PermanentAddrTTL)
+
+	log.Debug("Connected to Cumulus Peer:")
+	log.Debugf("Peer ID: %s", peerid.Pretty())
+	log.Debug("Peer Address:", targetAddr)
+
+	// Open a stream with the peer
+	stream, err := p.NewStream(context.Background(), peerid,
+		CumulusProtocol)
+	if err != nil {
+		return nil, err
 	}
+
+	mAddr, err := ma.NewMultiaddr(peerma)
+	if err != nil {
+		stream.Close()
+		return nil, err
+	}
+
+	err = p.subnet.AddPeer(mAddr, stream)
+	if err != nil {
+		stream.Close()
+		return nil, err
+	}
+
+	return stream, err
+}
+
+// Broadcast sends message to all peers this peer is currently connected to
+func (p *Peer) Broadcast(m msg.Message) error {
+	return errors.New("Function not implemented")
 }
 
 // ExtractPeerInfo extracts the peer ID and multiaddress from the
 // given multiaddress.
 // Returns peer ID (esentially 46 character hash created by the peer)
 // and the peer's multiaddress in the form /ip4/<peer IP>/tcp/<CumulusPort>.
```

Comment:
I'm not really sure that this function is necessary. I feel like it's doing much more work than it has to.